MediaBrowser.Controller/MediaEncoding/DownMixAlgorithmsHelper.cs

```diff
@@ -0,0 +1,61 @@
+using System.Collections.Generic;
+using MediaBrowser.Model.Entities;
+
+namespace MediaBrowser.Controller.MediaEncoding;
+
+/// <summary>
+/// Describes the downmix algorithms capabilities.
+/// </summary>
+public static class DownMixAlgorithmsHelper
+{
+    /// <summary>
+    /// The filter string of the DownMixStereoAlgorithms.
+    /// The index is the tuple of (algorithm, layout).
+    /// </summary>
+    public static readonly Dictionary<(DownMixStereoAlgorithms, string), string> AlgorithmFilterStrings = new()
+    {
+        { (DownMixStereoAlgorithms.Dave750, "5.1"), "pan=stereo|c0=0.5*c2+0.707*c0+0.707*c4+0.5*c3|c1=0.5*c2+0.707*c1+0.707*c5+0.5*c3" },
+        { (DownMixStereoAlgorithms.NightmodeDialogue, "5.1"), "pan=stereo|c0=c2+0.30*c0+0.30*c4|c1=c2+0.30*c1+0.30*c5" },
+        { (DownMixStereoAlgorithms.Rfc7845, "3.0"), "pan=stereo|c0=0.414214*c2+0.585786*c0|c1=0.414214*c2+0.585786*c1" },
+        { (DownMixStereoAlgorithms.Rfc7845, "quad"), "pan=stereo|c0=0.422650*c0+0.366025*c2+0.211325*c3|c1=0.422650*c1+0.366025*c3+0.211325*c2" },
+        { (DownMixStereoAlgorithms.Rfc7845, "5.0"), "pan=stereo|c0=0.460186*c2+0.650802*c0+0.563611*c3+0.325401*c4|c1=0.460186*c2+0.650802*c1+0.563611*c4+0.325401*c3" },
+        { (DownMixStereoAlgorithms.Rfc7845, "5.1"), "pan=stereo|c0=0.374107*c2+0.529067*c0+0.458186*c4+0.264534*c5+0.374107*c3|c1=0.374107*c2+0.529067*c1+0.458186*c5+0.264534*c4+0.374107*c3" },
+        { (DownMixStereoAlgorithms.Rfc7845, "6.1"), "pan=stereo|c0=0.321953*c2+0.455310*c0+0.394310*c5+0.227655*c6+0.278819*c4+0.321953*c3|c1=0.321953*c2+0.455310*c1+0.394310*c6+0.227655*c5+0.278819*c4+0.321953*c3" },
+        { (DownMixStereoAlgorithms.Rfc7845, "7.1"), "pan=stereo|c0=0.274804*c2+0.388631*c0+0.336565*c6+0.194316*c7+0.336565*c4+0.194316*c5+0.274804*c3|c1=0.274804*c2+0.388631*c1+0.336565*c7+0.194316*c6+0.336565*c5+0.194316*c4+0.274804*c3" },
+    };
+
+    /// <summary>
+    /// Get the audio channel layout string from the audio stream
+    /// If the input audio string does not have a valid layout string, guess from channel count.
+    /// </summary>
+    /// <param name="audioStream">The audio stream to get layout.</param>
+    /// <returns>Channel Layout string.</returns>
+    public static string InferChannelLayout(MediaStream audioStream)
+    {
+        if (!string.IsNullOrWhiteSpace(audioStream.ChannelLayout))
```

Comment:
`ChannelLayout` should be normalized before being used like this.
For BDMVs it will always be "$CHANNELS.$LFE" and never "stereo" or "quad"

Reply:
> For BDMVs it will always be "$CHANNELS.$LFE" and never "stereo" or "quad"

ffmpeg never outputs 2.0 as "2.0", it will always output stereo

Reply:
And "quad" and "4.0" are actually two different channel layouts. 4.0 means front left, front right ,front center and back center, but quad means front left, front right, back left and back right. All names have their own layout so it is not a normalization problem, it is a correctness problem.

Reply:
If our BDMV parsing code has such normalization, we should revert them. I hope we don't.

Reply:
> ffmpeg never outputs 2.0 as "2.0", it will always output stereo

That's bc we don't get that info from ffmpeg for BDMVs, we get it from the playlist info

Reply:
Just looked at the BDMV parser, now we have the problem of unable to identify the correct layout due to lack of information. I personally *want* to cowardly reject such ambiguous layouts and let ffmpeg's internal downmixing system to handle it as ffmpeg knows the layout better than us. The good news is that, we don't need to worry about the stereo case and the quad case is relatively rare compared to more common 5.1 or 7.1.

Reply:
I think we are safe for this case after investigation the current BD implementations. Although dolby digital technically supports 2/2 quad configuration but it is extremely rare and that layout is usually seen on legacy DVDs and music albums. So the 4.0 for BDs would almost always be the 4.0 in ffmpeg definition. I think we can just add a comment to the code just in case there is really an exception.

Reply:
The RFC7845 explicitly requires the quad (2 front 2 rear) layout to use that downmix algorithm, so for 4 channel BDMVs, we will always use ffmpeg's internal system which will be safe enough.